src/app/util/MatterCallbacks.h

```diff
@@ -14,32 +13,57 @@
  *    See the License for the specific language governing permissions and
  *    limitations under the License.
  */
-
-// THIS FILE IS GENERATED BY ZAP
-
 #pragma once
 
 #include <access/SubjectDescriptor.h>
 #include <app/ConcreteAttributePath.h>
 #include <app/ConcreteCommandPath.h>
 
-void MatterPreAttributeReadCallback(const chip::app::ConcreteAttributePath & attributePath);
-void MatterPostAttributeReadCallback(const chip::app::ConcreteAttributePath & attributePath);
-void MatterPreAttributeWriteCallback(const chip::app::ConcreteAttributePath & attributePath);
-void MatterPostAttributeWriteCallback(const chip::app::ConcreteAttributePath & attributePath);
+namespace chip {
 
-/** @brief Matter Pre Command Received
- *
- * This callback is called once the message has been determined to be a command, and
- * before the command is dispatched to the receiver.
- */
-CHIP_ERROR MatterPreCommandReceivedCallback(const chip::app::ConcreteCommandPath & commandPath,
-                                            const chip::Access::SubjectDescriptor & subjectDescriptor);
+/// Allows for application hooks for processing attribute and command operations
+class ApplicationCallbacks
+{
+public:
+    enum class OperationType
+    {
+        Read,
+        Write
+    };
 
-/** @brief Matter Post Command Received
- *
- * This callback is called once the message has been determined to be a command, but
- * after it being dispatched to the receiver.
- */
-void MatterPostCommandReceivedCallback(const chip::app::ConcreteCommandPath & commandPath,
-                                       const chip::Access::SubjectDescriptor & subjectDescriptor);
+    enum class OperationOrder
+    {
+        Pre,
+        Post
+    };
+
+    virtual ~ApplicationCallbacks() = default;
+
+    /// This callback is called on attribute operations:
+    /// - for reads and writes
+    /// - both befor and after attribute read/writes
+    ///
+    /// NOTE: PostRead is only called on read success.
+    virtual void AttributeOperation(OperationType operation, OperationOrder order, const chip::app::ConcreteAttributePath & path) {}
+
+    /// This callback is called once the message has been determined to be a command, and
+    /// before the command is dispatched to the receiver.
+    ///
+    /// Returning an error here will prevent the command to be dispatched further
+    virtual CHIP_ERROR PreCommandReceived(const chip::app::ConcreteCommandPath & commandPath,
```

Comment:
Is there a reason there's a single function for all attribute related events and two (post/pre) for commands?

Reply:
Pre-command allowed returning a CHIP_ERROR, so I kept that split.

Feedback of what would feel a good api would be appreciated though. The prewrite/postwrite/preread/postread felt duplicated, however I am unsure if the resulting bit is better.

Reply:
Also the `PostRead` being only called on success felt off, however I was not sure if that was intentional or not ... so I kept the comment and existing code. It feels wrong though as it is asymmetric with write.

Likely "post" should have a CHIP_ERROR attached to it with the state of the read/write. I could do that change too if that makes sense to others.

Reply:
Ah, makes sense, thanks!